tstrait/trait.py

```diff
@@ -0,0 +1,227 @@
+                elif current_state == causal_state:
+                    counts -= tree.num_samples(m.node)
+
+        return counts
+
+    def _frequency_multiplier(self, allele_freq):
+        """
+        Calculates the frequency dependence constant [2p(1-p)]^alpha by using
+        the allele_freq input.
+        """
+        if allele_freq == 0 or allele_freq == 1:
+            const = 0
+        else:
+            const = np.sqrt(pow(2 * allele_freq * (1 - allele_freq), self.alpha))
+        return const
+
+    def _freq_dep(self, site_id_array, causal_state_array):
+        """
+        Obtains the frequency dependent constant [2p(1-p)]^alpha for each site ID.
+        """
+        tree = tskit.Tree(self.ts)
+        num_samples = self.ts.num_samples
+        freq_dep_array = []
+        for site_id, causal_state in zip(site_id_array, causal_state_array):
+            site = self.ts.site(site_id)
+            tree.seek(site.position)
+            freq = self._obtain_allele_count(tree, site, causal_state) / num_samples
+            freq_dep_array.append(self._frequency_multiplier(freq))
+
+        return freq_dep_array
+
+    def _sim_beta(self, site_id_array, causal_state_array):
+        """
+        Simulates effect size by using the model given in the `model` input. If `alpha`
+        is non-zero, frequency dependence architecture is used.
+        """
+        beta_array = self.model._sim_effect_size(
+            num_causal=self.num_causal, rng=self.rng
+        )
+
+        if self.alpha == 0:
+            return np.transpose(beta_array)
+        else:
+            freq_dep = self._freq_dep(site_id_array, causal_state_array)
+            return np.multiply(np.transpose(beta_array), freq_dep)
+
+    def _run(self):
+        """
+        This method runs a simulation based on the input parameters. This method randomly
+        chooses causal sites and the corresponding causal state based on the `num_causal`
+        input. Afterwards, effect size of each causal site is simulated based on the
+        trait model given by the `model` input. If `alpha` is non-zero, frequency
+        dependence architecture is used.
+        """
+        site_id_array = self._choose_causal_site()
+        causal_state_array = []
+        for site_id in site_id_array:
+            causal_state_array.append(self._choose_causal_state(site_id))
+
+        beta_array = self._sim_beta(site_id_array, causal_state_array)
+
+        num_trait = self.model.num_trait
+
+        trait_df = pd.DataFrame(
+            {
+                "site_id": np.repeat(site_id_array, num_trait),
+                "effect_size": beta_array.flatten(order="F"),
+                "causal_state": np.repeat(causal_state_array, num_trait),
+                "trait_id": np.tile(np.arange(num_trait), self.num_causal),
+            }
+        )
+
+        return trait_df
+
+
+def sim_trait(ts, num_causal, model, alpha=0, random_seed=None):
```

Comment:
I think we should change the signature to:
```
def sim_trait(ts, model, *, num_causal=None, alpha=0, random_seed=None):
      num_causal = 1 if num_causal is None else num_causal
```

Basically, you have to define a ``ts`` and a ``model``, and there's no sensible defaults for this. It's easy to remember the pattern ``sim_trait(ts, model)``. It's quite reasonable to have a trait with 1 causal site, I think, so that doesn't have to be a mandatory argument. It also gives more flexibility in how we specify the causal sites later. 

You could imagine doing something like this later:
```
df = sim_trait(ts, model, causal=np.arange(ts.num_sites))
```
This doesn't work if ``num_causal`` is a required argument.

Reply:
I just made some changes, and I think that is a great idea!